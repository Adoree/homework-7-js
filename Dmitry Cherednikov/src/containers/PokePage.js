import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';

import FetchError from '../components/FetchError';
import Page from '../components/Page';
import { getPokemon, getIsFetchingPoke, getError } from '../reducers';
import { fetchPokemon } from '../actions';

class PokePage extends Component {
  static propTypes = {
    poke: PropTypes.object,
    isFetchingPoke: PropTypes.bool.isRequired,
  };

  static defaultProps = {
    fetchPokemon: () => {},
  };

  getPokemon = () => {
    if (!this.props.poke) {
      this.props.fetchPokemon(this.props.match.params.id);
    }
  };

  componentDidMount() {
    this.getPokemon()
  }

  render() {
    const { poke, isFetchingPoke, error } = this.props;
    if (isFetchingPoke) {
      return (
        <p>
          Fetching...
        </p>
      )
    }

    if (!poke && error) {
      return (
        <FetchError message={error} onRetry={this.getPokemon}/>
      )
    }

    return (
      <Page {...this.props.poke }/>
    )
  }
}

const mapStateToProps = (state, { match }) => {
  const id = match.params.id;
  return {
    poke: getPokemon(state, id),
    isFetchingPoke: getIsFetchingPoke(state),
    error: getError(state),
}};

const mapDispatchToProps = (dispatch) => ({
  fetchPokemon: bindActionCreators(fetchPokemon, dispatch)
});

PokePage = connect(
  mapStateToProps,
  mapDispatchToProps,
)(PokePage);

export default PokePage;