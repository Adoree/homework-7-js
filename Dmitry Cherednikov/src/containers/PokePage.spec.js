import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { PokePage } from './PokePage.js';

const setup = () => {
  const props = {
    poke: {
      name: 'pidgey',
      id: 16,
      date: '18.07.2018, 17:46:24',
    },
    isFetching: false,
    fetchPokemon: jest.fn(),
  };

  const output = shallow((
    <PokePage { ...props } />
  ));

  return {
    output,
    props,
  }
};

describe('<PokePage />', () => {
  it('renders correctly', () => {
    const { output } = setup();

    expect(toJson(output)).toMatchSnapshot();
  });

  it('renders FetchError component when error occurs if we have no pokemon in props', () => {
    const { output } = setup();
    const errorMessage = 'Something went wrong';

    expect(output.find('FetchError')).toHaveLength(0);

    output.setProps({ poke: null, errorMessage });

    expect(output.find('FetchError')).toHaveLength(1);
  });

  it('renders special message if there is no pokemon found', () => {
    const { output } = setup();
    const message = 'No pokemon found :(';

    expect(output.find('p')).toHaveLength(0);

    output.setProps({ poke: null });

    expect(output.find('p')).toHaveLength(1);
    expect(output.find('p').text()).toEqual(message);
  });

  it('renders special message if we are fetching data', () => {
    const { output } = setup();
    const message = 'Fetching...';

    expect(output.find('p')).toHaveLength(0);

    output.setProps({ poke: null, isFetching: true });

    expect(output.find('p')).toHaveLength(1);
    expect(output.find('p').text()).toEqual(message);
  })
});