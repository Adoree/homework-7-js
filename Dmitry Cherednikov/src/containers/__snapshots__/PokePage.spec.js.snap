// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokePage /> renders correctly 1`] = `
<Page
  date="18.07.2018, 17:46:24"
  id={16}
  name="pidgey"
  onBack={[Function]}
/>
`;
